lib/evm_database_ops.rb

```diff
@@ -115,14 +116,18 @@ def self.restore(db_opts, connect_opts = {})
         uri = File.join(connect_opts[:uri], backup_folder, connect_opts[:remote_file_name])
       end
 
+      connect_opts[:username] = db_opts[:username] if connect_opts[:username].nil?
+      connect_opts[:password] = db_opts[:password] if connect_opts[:password].nil?
       session = MiqGenericMountSession.new_session(connect_opts)
-      db_opts[:local_file] = session.uri_to_local_path(uri)
+      db_opts[:local_file]  = session.uri_to_local_path(uri)
+      db_opts[:object_file] = session.uri_to_object_path(uri) if connect_opts[:object_store]
     end
 
     block_result = yield db_opts if block_given?
     uri || block_result
   ensure
-    session.disconnect if session
+    session&.add(db_opts[:local_file], connect_opts[:uri], db_opts[:object_file]) if action == :backup && session.class.name == "MiqS3Session"
```

Comment:
I don't think this belongs in an `ensure` block as if an error occurred there's no guarantee that a backup was produced.